### ISSUE TITLE
Config watcher not sending HTTP request after 404

### ISSUE THREAD

**Issue by arobie1992** [spring-cloud/spring-cloud-kubernetes#831]
I'm doing a poc of the spring config watcher. We set up the watcher with the yaml provided [here](https://docs.spring.io/spring-cloud-kubernetes/docs/current/reference/html/index.html#deployment-yaml) and are using an nginx pod as the test service that the watcher should ping to refresh. 

The first time the config map gets updated, the watcher logs show it sending the post request to the actuator endpoint and getting a 404 back. That's expected, but subsequent requests appear to not get sent at all. If I restart the config watcher, then it does send the request again, but if I remove the target app's pod and create a new one, it still doesn't make the post request.

It seems like it might be blacklisting that URL, which would make sense so the config watcher isn't constantly sending requests to an endpoint that doesn't exist, but I didn't see anything that seemed like it would be doing that. So I was wondering what the cause of this behavior is and if there's a way to trigger the watcher to send posts to the endpoint again without restarting it.

**Comment by ryanjbaxter** [comment]
There certainly is no kind of black listing happening in the config watcher.  Can you provide a repo with a deployment for nginx pod and the config map you created?

**Comment by arobie1992** [comment]
Here's the yaml for everything we've been testing with: https://github.com/arobie1992/spring-config-watcher-demo/blob/main/demo-config-watcher.yaml

I did make a few minor tweaks to point straight to docker hub for the images rather than proxying through our image repo—it just goes to docker hub for the initial download anyway. Let me know if you need anything further.

**Comment by ryanjbaxter** [comment]
How are you testing that the configuration has been changed in the NGINX pod?

**Comment by arobie1992** [comment]
We were just shelling into the pod and catting the contents of `/demo-config/demo-config` to confirm that the config map got updated. Nothing as far as the app itself. This setup was just to make sure we could set up the config watcher and that it sent the post request.

**Comment by ryanjbaxter** [comment]
The only thing the config watcher does is notify the app via http or via spring cloud bus that configuration they are interested in has been updated.  It has nothing to do with whether the configuration that is mounted in the pod get updated.  So if you are catting the config and not seeing it updated that is not an issue with Spring Cloud Kubernetes, Kubernetes itself should be updating the mounted config map.

**Comment by arobie1992** [comment]
We are seeing the config map get updated. The behavior that I had a question about is that it doesn't seem to send the HTTP request after the initial 404 failure (including what seem to be a couple retries). So the order of events is about as follows.

1. Apply the linked yaml to the cluster and wait for the pods to start
2. Make a change to the config map
3. Config watcher sends HTTP request, gets 404, and logs stack traces
4. Wait for a bit to ensure no more 404 stack traces get logged from previous request
5. Make another change to the config map
6. No sign of 404 responses from request

 I can post some logs if that would help.

**Comment by ryanjbaxter** [comment]
I am not seeing that when I deploy the apps locally, here you can see 2 404 logs from the watcher calling the NGINX app

```
2021/07/19 14:35:17 [error] 32#32: *1 open() "/usr/share/nginx/html/actuator/refresh" failed (2 │
│ 10.244.0.5 - - [19/Jul/2021:14:35:17 +0000] "POST /actuator/refresh HTTP/1.1" 404 153 "-" "Reac │
│ 2021/07/19 14:39:18 [error] 32#32: *2 open() "/usr/share/nginx/html/actuator/refresh" failed (2 │
│ 10.244.0.5 - - [19/Jul/2021:14:39:18 +0000] "POST /actuator/refresh HTTP/1.1" 404 153 "-" "Reac │
```

**Comment by spring-cloud-issues** [comment]
If you would like us to look at this issue, please provide the requested information. If the information is not provided within the next 7 days this issue will be closed.

**Comment by arobie1992** [comment]
Did a bit more troubleshooting today. Turns out it's not caused by the 404 response but instead has to do with leaving it idle for a while; I tested with about 30 minutes. I have some logs and notes of when I made changes that I can upload in a bit if it would be helpful—mostly they just need cleaning up. I'm working on setting up a small pod that will return 200 to see how that behaves and will post here when I have an update.

**Comment by ryanjbaxter** [comment]
Yes that would be helpful!

**Comment by arobie1992** [comment]
[spring-config-watcher.zip](https://github.com/spring-cloud/spring-cloud-kubernetes/files/6888708/spring-config-watcher.zip)

Here's my logs and the records of what changes I made and approximately when. At one point, I thought the value might matter, so those are in there as well, but you can probably ignore it. They should typically have my computer's time as well as the time from the logs. Let me know if you have any questions about them.

**Comment by ryanjbaxter** [comment]
Fixed via 888031d
